In [3]:
from pyspark.sql import SparkSession
from pyspark.sql import functions as F
from pyspark.ml.feature import StringIndexer, OneHotEncoder, VectorAssembler
from pyspark.ml.regression import GBTRegressor
from pyspark.ml import Pipeline
from pyspark.ml.evaluation import RegressionEvaluator

spark = SparkSession.builder \
    .appName("Taxi ETL") \
    .config("spark.jars.packages", "org.postgresql:postgresql:42.6.0") \
    .getOrCreate()

In [4]:
df=spark.read.parquet('../data/dataset.parquet')
df.show(5)


+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+------------------+
|VendorID|tpep_pickup_datetime|tpep_dropoff_datetime|passenger_count|trip_distance|RatecodeID|store_and_fwd_flag|PULocationID|DOLocationID|payment_type|fare_amount|extra|mta_tax|tip_amount|tolls_amount|improvement_surcharge|total_amount|congestion_surcharge|Airport_fee|cbd_congestion_fee|
+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+------------------+
|       1| 2025-01-01 00:18:38|  2025-01-01 00:26:59|              1|          1.6|         1|                 N|         229|    

In [5]:
df.printSchema()

root
 |-- VendorID: integer (nullable = true)
 |-- tpep_pickup_datetime: timestamp_ntz (nullable = true)
 |-- tpep_dropoff_datetime: timestamp_ntz (nullable = true)
 |-- passenger_count: long (nullable = true)
 |-- trip_distance: double (nullable = true)
 |-- RatecodeID: long (nullable = true)
 |-- store_and_fwd_flag: string (nullable = true)
 |-- PULocationID: integer (nullable = true)
 |-- DOLocationID: integer (nullable = true)
 |-- payment_type: long (nullable = true)
 |-- fare_amount: double (nullable = true)
 |-- extra: double (nullable = true)
 |-- mta_tax: double (nullable = true)
 |-- tip_amount: double (nullable = true)
 |-- tolls_amount: double (nullable = true)
 |-- improvement_surcharge: double (nullable = true)
 |-- total_amount: double (nullable = true)
 |-- congestion_surcharge: double (nullable = true)
 |-- Airport_fee: double (nullable = true)
 |-- cbd_congestion_fee: double (nullable = true)



In [6]:
df.show()

+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+------------------+
|VendorID|tpep_pickup_datetime|tpep_dropoff_datetime|passenger_count|trip_distance|RatecodeID|store_and_fwd_flag|PULocationID|DOLocationID|payment_type|fare_amount|extra|mta_tax|tip_amount|tolls_amount|improvement_surcharge|total_amount|congestion_surcharge|Airport_fee|cbd_congestion_fee|
+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+------------------+
|       1| 2025-01-01 00:18:38|  2025-01-01 00:26:59|              1|          1.6|         1|                 N|         229|    

In [7]:
df.columns

['VendorID',
 'tpep_pickup_datetime',
 'tpep_dropoff_datetime',
 'passenger_count',
 'trip_distance',
 'RatecodeID',
 'store_and_fwd_flag',
 'PULocationID',
 'DOLocationID',
 'payment_type',
 'fare_amount',
 'extra',
 'mta_tax',
 'tip_amount',
 'tolls_amount',
 'improvement_surcharge',
 'total_amount',
 'congestion_surcharge',
 'Airport_fee',
 'cbd_congestion_fee']

In [8]:
df.dtypes

[('VendorID', 'int'),
 ('tpep_pickup_datetime', 'timestamp_ntz'),
 ('tpep_dropoff_datetime', 'timestamp_ntz'),
 ('passenger_count', 'bigint'),
 ('trip_distance', 'double'),
 ('RatecodeID', 'bigint'),
 ('store_and_fwd_flag', 'string'),
 ('PULocationID', 'int'),
 ('DOLocationID', 'int'),
 ('payment_type', 'bigint'),
 ('fare_amount', 'double'),
 ('extra', 'double'),
 ('mta_tax', 'double'),
 ('tip_amount', 'double'),
 ('tolls_amount', 'double'),
 ('improvement_surcharge', 'double'),
 ('total_amount', 'double'),
 ('congestion_surcharge', 'double'),
 ('Airport_fee', 'double'),
 ('cbd_congestion_fee', 'double')]

In [9]:
df.describe().show()

26/01/12 14:20:49 WARN SparkStringUtils: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.


+-------+------------------+------------------+-----------------+-----------------+------------------+-----------------+------------------+------------------+------------------+------------------+-------------------+------------------+-------------------+---------------------+------------------+--------------------+-------------------+-------------------+
|summary|          VendorID|   passenger_count|    trip_distance|       RatecodeID|store_and_fwd_flag|     PULocationID|      DOLocationID|      payment_type|       fare_amount|             extra|            mta_tax|        tip_amount|       tolls_amount|improvement_surcharge|      total_amount|congestion_surcharge|        Airport_fee| cbd_congestion_fee|
+-------+------------------+------------------+-----------------+-----------------+------------------+-----------------+------------------+------------------+------------------+------------------+-------------------+------------------+-------------------+---------------------+-------

In [10]:
from pyspark.sql.functions import col, count, when

df.select([
    count(when(col(c).isNull(), c)).alias(c)
    for c in df.columns
]).show()


+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+------------------+
|VendorID|tpep_pickup_datetime|tpep_dropoff_datetime|passenger_count|trip_distance|RatecodeID|store_and_fwd_flag|PULocationID|DOLocationID|payment_type|fare_amount|extra|mta_tax|tip_amount|tolls_amount|improvement_surcharge|total_amount|congestion_surcharge|Airport_fee|cbd_congestion_fee|
+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+------------------+
|       0|                   0|                    0|         540149|            0|    540149|            540149|           0|    

In [11]:
df_clean = df.na.drop()


In [12]:
df_clean.select([
    count(when(col(c).isNull(), c)).alias(c)
    for c in df.columns
]).show()


+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+------------------+
|VendorID|tpep_pickup_datetime|tpep_dropoff_datetime|passenger_count|trip_distance|RatecodeID|store_and_fwd_flag|PULocationID|DOLocationID|payment_type|fare_amount|extra|mta_tax|tip_amount|tolls_amount|improvement_surcharge|total_amount|congestion_surcharge|Airport_fee|cbd_congestion_fee|
+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+------------------+
|       0|                   0|                    0|              0|            0|         0|                 0|           0|    

In [13]:
from pyspark.sql.functions import unix_timestamp

# Convertir les timestamps en secondes et faire la différence
df_clean = df_clean.withColumn(
    "trip_duration",
    (unix_timestamp(col("tpep_dropoff_datetime")) - unix_timestamp(col("tpep_pickup_datetime"))) / 60
)

# df_clean.select("tpep_pickup_datetime", "tpep_dropoff_datetime", "trip_duration").show()
# Afficher les lignes avec trip_duration <= 0
df_clean.filter(col("trip_duration") <= 0).count()

1832

In [14]:
df_clean = df_clean.filter(col("trip_duration") > 0)
df_clean.show()


+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+------------------+------------------+
|VendorID|tpep_pickup_datetime|tpep_dropoff_datetime|passenger_count|trip_distance|RatecodeID|store_and_fwd_flag|PULocationID|DOLocationID|payment_type|fare_amount|extra|mta_tax|tip_amount|tolls_amount|improvement_surcharge|total_amount|congestion_surcharge|Airport_fee|cbd_congestion_fee|     trip_duration|
+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+------------------+------------------+
|       1| 2025-01-01 00:18:38|  2025-01-01 00:26:59|              1|    

In [15]:
df_clean = df_clean.filter((col("passenger_count").isNotNull()) & (col("passenger_count") > 0))
df_clean.show()


+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+------------------+------------------+
|VendorID|tpep_pickup_datetime|tpep_dropoff_datetime|passenger_count|trip_distance|RatecodeID|store_and_fwd_flag|PULocationID|DOLocationID|payment_type|fare_amount|extra|mta_tax|tip_amount|tolls_amount|improvement_surcharge|total_amount|congestion_surcharge|Airport_fee|cbd_congestion_fee|     trip_duration|
+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+------------------+------------------+
|       1| 2025-01-01 00:18:38|  2025-01-01 00:26:59|              1|    

In [16]:
df_clean = df_clean.drop("duration")

df_clean.show(10)


+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+------------------+------------------+
|VendorID|tpep_pickup_datetime|tpep_dropoff_datetime|passenger_count|trip_distance|RatecodeID|store_and_fwd_flag|PULocationID|DOLocationID|payment_type|fare_amount|extra|mta_tax|tip_amount|tolls_amount|improvement_surcharge|total_amount|congestion_surcharge|Airport_fee|cbd_congestion_fee|     trip_duration|
+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+------------------+------------------+
|       1| 2025-01-01 00:18:38|  2025-01-01 00:26:59|              1|    

In [17]:
# Verification du nombre de colonnes après nettoyage
len(df_clean.columns)


21

#### Identification des colonnes de types de données

In [18]:
#Get All column names and it's types
for col in df_clean.dtypes:
    print(col[0]+" , "+col[1])

VendorID , int
tpep_pickup_datetime , timestamp_ntz
tpep_dropoff_datetime , timestamp_ntz
passenger_count , bigint
trip_distance , double
RatecodeID , bigint
store_and_fwd_flag , string
PULocationID , int
DOLocationID , int
payment_type , bigint
fare_amount , double
extra , double
mta_tax , double
tip_amount , double
tolls_amount , double
improvement_surcharge , double
total_amount , double
congestion_surcharge , double
Airport_fee , double
cbd_congestion_fee , double
trip_duration , double


In [19]:
#store all column names in a list
allCols = [item[0] for item in df_clean.dtypes]

In [20]:
# Colonnes catégorielles : string + certaines colonnes bigint que tu sais être des catégories
categoricalCols = [c for c, t in df_clean.dtypes if t.startswith('string')] + ["payment_type", "RatecodeID", "VendorID"]
print(categoricalCols)

['store_and_fwd_flag', 'payment_type', 'RatecodeID', 'VendorID']


In [21]:
# Colonnes continues : int/bigint/double qui sont des nombres réels à calculer
numsCols = [c for c, t in df_clean.dtypes if t in ('double','int','bigint') and c not in categoricalCols]
print(numsCols)

['passenger_count', 'trip_distance', 'PULocationID', 'DOLocationID', 'fare_amount', 'extra', 'mta_tax', 'tip_amount', 'tolls_amount', 'improvement_surcharge', 'total_amount', 'congestion_surcharge', 'Airport_fee', 'cbd_congestion_fee', 'trip_duration']


In [22]:
from pyspark.sql import functions as F
from functools import reduce

def find_outliers(df, numerical_columns=None):
    """
    Ajoute des colonnes pour marquer les outliers et compte le total
    """
    if numerical_columns is None:
        numeric_columns = [c for c,t in df.dtypes if t in ('int','bigint','double')]
    else:
        numeric_columns = numerical_columns

    for column in numeric_columns:
        Q1 = df.approxQuantile(column,[0.25],0)[0]
        Q3 = df.approxQuantile(column,[0.75],0)[0]
        IQR = Q3 - Q1

        lower = Q1 - 1.5*IQR
        upper = Q3 + 1.5*IQR

        df = df.withColumn(f'is_outlier_{column}', F.when((F.col(column) < lower) | (F.col(column) > upper), 1).otherwise(0))

    selected_columns = [c for c in df.columns if c.startswith("is_outlier")]
    df = df.withColumn('total_outliers', reduce(lambda a,b: a+b, [F.col(c) for c in selected_columns]))
    df = df.drop(*selected_columns)

    return df

- Les colonnes is_outlier_* sont supprimées pour garder un dataframe propre, mais total_outliers reste.
- Utilisation de la fonction Outlier personnalisée et application à une trame de données Spark


In [23]:
df_clean = find_outliers(df_clean, numsCols)
df_clean.show()

+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+------------------+------------------+--------------+
|VendorID|tpep_pickup_datetime|tpep_dropoff_datetime|passenger_count|trip_distance|RatecodeID|store_and_fwd_flag|PULocationID|DOLocationID|payment_type|fare_amount|extra|mta_tax|tip_amount|tolls_amount|improvement_surcharge|total_amount|congestion_surcharge|Airport_fee|cbd_congestion_fee|     trip_duration|total_outliers|
+--------+--------------------+---------------------+---------------+-------------+----------+------------------+------------+------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+--------------------+-----------+------------------+------------------+--------------+
|       1| 2025-01-01 00:18:

In [24]:
from pyspark.sql import functions as F

df_silver = df_clean.filter(
    (F.col("trip_duration") > 0) &
    (F.col("trip_distance") > 0) &
    (F.col("trip_distance") <= 200) &
    (F.col("passenger_count") > 0) &
    (F.col("passenger_count") <= 8) & 
    (F.col("fare_amount") >= 0) &
    (F.col("extra") >= 0) &
    (F.col("mta_tax") >= 0) &
    (F.col("tip_amount") >= 0) &
    (F.col("tolls_amount") >= 0) &
    (F.col("improvement_surcharge") >= 0) &
    (F.col("total_amount") >= 0) &
    (F.col("congestion_surcharge") >= 0) &
    (F.col("Airport_fee") >= 0) &
    (F.col("cbd_congestion_fee") >= 0)
)


Ajustement de la trame de données ci-dessus, pour sélectionner uniquement les enregistrements où le nombre de valeurs aberrantes est < = 1

In [25]:

print("\n=== DÉTECTION DES OUTLIERS (IQR) ===")
df_with_outliers = find_outliers(df_clean, numsCols)

# Voir la distribution des outliers
print("\n=== DISTRIBUTION DES OUTLIERS PAR LIGNE ===")
df_with_outliers.groupBy('total_outliers').count().orderBy('total_outliers').show()




=== DÉTECTION DES OUTLIERS (IQR) ===



=== DISTRIBUTION DES OUTLIERS PAR LIGNE ===


+--------------+-------+
|total_outliers|  count|
+--------------+-------+
|             0|1887285|
|             1| 545734|
|             2|  57229|
|             3|  40099|
|             4|  53625|
|             5|  95551|
|             6| 103717|
|             7|  85609|
|             8|  31689|
|             9|   6548|
|            10|   1482|
|            11|    211|
|            12|      1|
+--------------+-------+



In [26]:
df_silver = df_with_outliers.filter(F.col('total_outliers') <= 4)

In [27]:
df_silver.count()

2583972

Converting a spark dataframe into pandas dataframe

In [28]:
df_silver.printSchema()



root
 |-- VendorID: integer (nullable = true)
 |-- tpep_pickup_datetime: timestamp_ntz (nullable = true)
 |-- tpep_dropoff_datetime: timestamp_ntz (nullable = true)
 |-- passenger_count: long (nullable = true)
 |-- trip_distance: double (nullable = true)
 |-- RatecodeID: long (nullable = true)
 |-- store_and_fwd_flag: string (nullable = true)
 |-- PULocationID: integer (nullable = true)
 |-- DOLocationID: integer (nullable = true)
 |-- payment_type: long (nullable = true)
 |-- fare_amount: double (nullable = true)
 |-- extra: double (nullable = true)
 |-- mta_tax: double (nullable = true)
 |-- tip_amount: double (nullable = true)
 |-- tolls_amount: double (nullable = true)
 |-- improvement_surcharge: double (nullable = true)
 |-- total_amount: double (nullable = true)
 |-- congestion_surcharge: double (nullable = true)
 |-- Airport_fee: double (nullable = true)
 |-- cbd_congestion_fee: double (nullable = true)
 |-- trip_duration: double (nullable = true)
 |-- total_outliers: integer (n

In [29]:
# from sqlalchemy import create_engine


# DB_HOST="localhost"
# DB_PORT="5432"
# DB_NAME="taxi_nyc"
# DB_USER="taxi_user"
# DB_PASSWORD="manal.2480"

# engine = create_engine(
#     f"postgresql://{DB_USER}:{DB_PASSWORD}"
#     f"@{DB_HOST}:{DB_PORT}/{DB_NAME}"
# )

# df_test = df_silver.toPandas()
# df_test.to_sql("taxi_trips", engine, if_exists="replace", index=False)



L'ensemble de données, qui contient 2 valeurs aberrantes ou plus dans chaque enregistrement


In [30]:
data_with_outliers = df_silver.filter(df_silver['total_Outliers']>=2)
data_with_outliers.count()

150953

In [31]:
# # Selecting the numerical columns from the original dataframe and converting into pandas
numsCols

['passenger_count',
 'trip_distance',
 'PULocationID',
 'DOLocationID',
 'fare_amount',
 'extra',
 'mta_tax',
 'tip_amount',
 'tolls_amount',
 'improvement_surcharge',
 'total_amount',
 'congestion_surcharge',
 'Airport_fee',
 'cbd_congestion_fee',
 'trip_duration']

In [32]:
# # Test de connexion simple
# import psycopg2

# try:
#     conn = psycopg2.connect(
#         host="postgres",
#         port=5432,
#         database="taxi",
#         user="taxi",
#         password="taxi"
#     )
#     print("✓ Connexion PostgreSQL réussie!")
#     conn.close()
# except Exception as e:
#     print(f"✗ Erreur de connexion: {e}")

In [33]:
# # Étape 3 : Sauvegarde PostgreSQL

# df.write \
#     .format("jdbc") \
#     .option("url", "jdbc:postgresql://postgres:5432/taxi") \
#     .option("dbtable", "silver_taxi_trips") \
#     .option("user", "taxi") \
#     .option("password", "taxi") \
#     .option("driver", "org.postgresql.Driver") \
#     .mode("overwrite") \
#     .save()

# print("✓ Données sauvegardées avec succès dans PostgreSQL!")



#### PRÉPARATION DES DONNÉES

In [34]:
from pyspark.ml.feature import StringIndexer, OneHotEncoder, VectorAssembler, UnivariateFeatureSelector
from pyspark.ml.regression import GBTRegressor
from pyspark.ml import Pipeline
from pyspark.ml.evaluation import RegressionEvaluator

# ===== 1. FEATURES TEMPORELLES  =====

df_prepared = df_silver \
    .withColumn("pickuphour", F.hour("tpep_pickup_datetime")) \
    .withColumn("dayof_week", F.dayofweek("tpep_pickup_datetime")) \
    .withColumn("speed", F.col("trip_distance") / (F.col("trip_duration") / 60))


print(f"Avant filtrage: {df_prepared.count():,} lignes")

# ===== 2. FILTRAGE STRICT =====
print("\nFiltrage des valeurs aberrantes...")

df_clean = df_prepared.filter(
    (F.col("trip_distance") > 0.1) &
    (F.col("trip_distance") < 50) &
    (F.col("trip_duration") > 1) &
    (F.col("trip_duration") < 120) &
    (F.col("passenger_count") > 0) &
    (F.col("passenger_count") <= 6) &
    (F.col("speed") > 1) &          
    (F.col("speed") < 60)          
)


# ===== 3. PLUS DE FEATURES =====
num_cols = [
    "trip_distance",
    "fare_amount",
    "tip_amount",
    "tolls_amount",
    "total_amount",
    "Airport_fee",
    "passenger_count",      
    "extra",                
    "mta_tax",              
    "congestion_surcharge", 
    "speed"                 
]

cat_cols = [
    "RatecodeID",
    "pickuphour",
    "dayof_week",
    "VendorID",        
    "PULocationID",    
    "DOLocationID",    
    "payment_type"     
]

target = "trip_duration"

print(f"\nFeatures numériques: {len(num_cols)}")
print(f"Features catégorielles: {len(cat_cols)}")

# ===== 4. SPLIT TRAIN/TEST =====
train_df, test_df = df_clean.randomSplit([0.8, 0.2], seed=42)

print(f"\nTrain: {train_df.count():,} lignes")
print(f"Test:  {test_df.count():,} lignes")


Avant filtrage: 2,583,972 lignes

Filtrage des valeurs aberrantes...

Features numériques: 11
Features catégorielles: 7



Train: 2,031,809 lignes


Test:  507,371 lignes


#### PIPELINE ML

In [35]:
# ===== 5. PIPELINE ML =====
indexers = [
    StringIndexer(inputCol=c, outputCol=f"{c}_idx", handleInvalid="keep")
    for c in cat_cols
]

encoders = [
    OneHotEncoder(inputCol=f"{c}_idx", outputCol=f"{c}_ohe")
    for c in cat_cols
]

feature_cols = num_cols + [f"{c}_ohe" for c in cat_cols]
assembler = VectorAssembler(
    inputCols=feature_cols,
    outputCol="features",
    handleInvalid="skip"
)

gbt = GBTRegressor(
    featuresCol="features",
    labelCol=target,
    maxIter=100,     
    maxDepth=6,      
    seed=42
)

pipeline = Pipeline(stages=indexers + encoders + [assembler, gbt])

#### ENTRAÎNEMENT

In [36]:
# ===== 6. ENTRAÎNEMENT =====

pipeline_model = pipeline.fit(train_df)


26/01/12 14:29:14 WARN MemoryStore: Not enough space to cache rdd_682_10 in memory! (computed 234.4 MiB so far)
26/01/12 14:29:14 WARN BlockManager: Persisting block rdd_682_10 to disk instead.
26/01/12 14:29:16 WARN MemoryStore: Not enough space to cache rdd_682_6 in memory! (computed 155.5 MiB so far)
26/01/12 14:29:16 WARN BlockManager: Persisting block rdd_682_6 to disk instead.
26/01/12 14:29:18 WARN MemoryStore: Not enough space to cache rdd_682_2 in memory! (computed 234.4 MiB so far)
26/01/12 14:29:18 WARN BlockManager: Persisting block rdd_682_2 to disk instead.
26/01/12 14:29:25 WARN MemoryStore: Not enough space to cache rdd_682_10 in memory! (computed 234.4 MiB so far)
26/01/12 14:29:25 WARN MemoryStore: Not enough space to cache rdd_682_10 in memory! (computed 43.7 MiB so far)
26/01/12 14:29:30 WARN MemoryStore: Not enough space to cache rdd_682_2 in memory! (computed 65.6 MiB so far)
26/01/12 14:29:31 WARN MemoryStore: Not enough space to cache rdd_682_2 in memory! (compu

#### PRÉDICTIONS

In [37]:
# ===== 7. PRÉDICTIONS =====
predictions = pipeline_model.transform(test_df)

#### ÉVALUATION

In [38]:
# from pyspark.sql import SparkSession

# spark = SparkSession.builder \
#     .master("local[*]") \
#     .appName("SparkOK") \
#     .getOrCreate()

# spark.range(10).show()

# ===== 8. ÉVALUATION =====
evaluator = RegressionEvaluator(labelCol=target, predictionCol="prediction")

rmse = evaluator.evaluate(predictions, {evaluator.metricName: "rmse"})
mae = evaluator.evaluate(predictions, {evaluator.metricName: "mae"})
r2 = evaluator.evaluate(predictions, {evaluator.metricName: "r2"})

print("\n" + "="*60)
print(f"RMSE: {rmse:.2f} minutes")
print(f"MAE:  {mae:.2f} minutes")
print(f"R²:   {r2:.4f} ")
print("="*60)



RMSE: 1.14 minutes
MAE:  0.51 minutes
R²:   0.9769 


#### SAUVEGARDER LE MODÈLE

In [40]:
# ===== 9. SAUVEGARDER LE MODÈLE =====

pipeline_model.write().overwrite().save("./models/eta_model")


J'ai amélioré mon R² de 0.89 à 0.98 grâce à trois changements principaux :

1. Ajout des zones géographiques (PULocationID/DOLocationID) - car la localisation impacte énormément la durée en ville
2. Création d'une feature vitesse calculée comme distance/temps - directement corrélée à notre cible
3. Nettoyage strict des outliers - pour avoir des données de meilleure qualité

Ces améliorations ont réduit mon erreur de 2.43 minutes à seulement 1.14 minutes.